dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -0,0 +1,742 @@
+    // (e.g. EXCLUDE CURRENT ROW), so we have to check both.
+    assert(prev_frame_start <= frame_start);
+    const auto first_used_block = std::min(next_output_block_number,
+                                           std::min(prev_frame_start.block, current_row.block));
+
+    if (first_block_number < first_used_block)
+    {
+        window_blocks.erase(window_blocks.begin(),
+                            window_blocks.begin() + (first_used_block - first_block_number));
+        first_block_number = first_used_block;
+
+        assert(next_output_block_number >= first_block_number);
+        assert(frame_start.block >= first_block_number);
+        assert(prev_frame_start.block >= first_block_number);
+        assert(current_row.block >= first_block_number);
+        assert(peer_group_start.block >= first_block_number);
+    }
+}
+
+void WindowBlockInputStream::appendBlock(Block & current_block)
+{
+    if (!input_is_finished)
+    {
+        if (current_block.rows() == 0)
+        {
+            return;
+        }
+
+        window_blocks.push_back({});
+        auto & window_block = window_blocks.back();
+
+        window_block.rows = current_block.rows();
+
+        // Initialize output columns and add new columns to output block.
+        for (auto & ws : workspaces)
+        {
+            ws.result = current_block.columns();
+            if (ws.is_agg_workspace)
+            {
+                throw Exception(ErrorCodes::NOT_IMPLEMENTED,
+                                "agg functions in window are not implemented");
+            }
+            else
+            {
+                window_block.output_columns.push_back(ws.window_function->getReturnType()->createColumn());
+            }
+        }
+
+        window_block.input_columns = current_block.getColumns();
+    }
+
+    // Start the calculations. First, advance the partition end.
+    for (;;)
+    {
+        advancePartitionEnd();
+
+        // Either we ran out of data or we found the end of partition (maybe
+        // both, but this only happens at the total end of data).
+        assert(partition_ended || partition_end == blocksEnd());
+        if (partition_ended && partition_end == blocksEnd())
+        {
+            assert(input_is_finished);
+        }
+
+        // if window only have row_number function, we can ignore frame.
+        while (current_row < partition_end)
+        {
+            if (!only_have_row_number)
+            {
+                // We now know that the current row is valid, so we can update the
+                // peer group start.
+                if (!arePeers(peer_group_start, current_row))
+                {
+                    peer_group_start = current_row;
+                    peer_group_start_row_number = current_row_number;
+                    ++peer_group_number;
+                }
+            }
+
+            // Advance the frame start.
+            advanceFrameStart();
+
+            if (!frame_started)
```

Comment:
That frame_started must be true is correct in terms of currently supported functions(row_number, dense_rank, rank), because these functions frame must be [current_row, current_row + 1), But the frame_start row in agg functions may be in the next block, on this occasion frame_started could be false.
I suggest no change here for later extension of agg functions.

Reply:
ok